Rating view

GeorgeLeithead opened the pull request:
### Description of Change ###
This pull request introduces a custom `RatingView` control, designed to provide developers with a flexible and customizable rating mechanism, similar to those used on popular review and feedback platforms. The control is versatile and adaptable, allowing developers to tailor it to the specific needs of their applications through a variety of customizable properties.

The primary goal of this control is to enhance the user experience by enabling precise and intuitive user ratings, while offering developers full control over the visual and functional aspects of the control. The following are key features and properties available in the `RatingView` control:

**Key Features and Property Descriptions:**
- Set the Maximum Number of rating items:
Developers can define the total number of items (e.g., stars, hearts, etc., or custom shapes) available for rating. This allows for ratings of any scale, such as a 5-star or 10-star system, depending on the needs of the application.

- Set the current Rating:
The control supports setting the current rating value, allowing for both pre-defined ratings (e.g., from previous user input or stored data) and updates during runtime as the user interacts with the control.

- Set the Filled (rated) background colour:
Developers can set the colour that will be applied to the filled (rated) portion of each item, offering flexibility in defining the visual aesthetic of the rating items when selected by the user.

- Set the Empty rating colour:
The colour for the unfilled (empty) rating items can also be customized. This allows for clear visual differentiation between rated and unrated items.

- Set the rating item shape Border Colour and Thickness:
The control allows for defining the border colour and thickness of each rating item. This provides additional flexibility to create visually distinct and stylized rating shapes with custom borders.

- Set the rating Fill type (Shape or Item):
Developers can choose how the fill is applied against the entire rating item or just the shape, enabling more nuanced visual presentation, such as filling only the interior of stars or the full item.

- Set the rating item Shape:
The control provides the ability to define the shape of the rating items, such as stars, circles, like, dislike, or any other commonly used rating icons.

- Define a Custom rating item shape:
In addition to standard shapes, the control allows for defining custom shapes for rating item shapes. This feature empowers developers to implement unique designs, such as distinctive symbols, as rating items.

- Set the rating item shape Size:
The size of the rating item shape can be customized to fit the overall design of the application, providing the flexibility to adjust the control to various UI layouts.

- Set the Spacing between rating items:
Developers can define the spacing between individual rating items, ensuring proper alignment and spacing based on the control’s context within the application's layout.

- Set the Padding between the rating item and the rating item shape:
Padding between the rating item and its corresponding shape can be adjusted, allowing for finer control over the appearance and layout of the rating items.

- Attach a Handler to the `RatingChanged` Event:
The control supports event handling through the `RatingChanged` event, enabling developers to respond to changes in the user’s rating input, such as updating data models, triggering animations, or submitting ratings to a backend service.

- Define the control in XAML or C# syntax:
The `RatingView` control can be defined and implemented using either XAML or C# syntax, making it highly versatile for both declarative and programmatic UI design, depending on developer preference or application architecture.

 ### Linked Issues ###

 - Fixes #1607 

 ### PR Checklist ###
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: [https://github.com/MicrosoftDocs/CommunityToolkit/pull/472](https://github.com/MicrosoftDocs/CommunityToolkit/pull/472)


 ### Additional information ###
This pull request also includes three sample pages to assist developers in exploring and implementing the `RatingView` control:

- XAML Syntax Sample Page:
Demonstrates the usage and configuration of the `RatingView` control using XAML, providing an example for developers who prefer a declarative approach to UI development.

- C# Syntax Sample Page:
Shows how the `RatingView` control can be implemented programmatically in C#, catering to developers who prefer to work in code for dynamic or logic-driven UIs.

- Showcase Page:
Provides a demonstration of various customization options and use cases for the `RatingView` control, offering developers inspiration on how they can integrate and style the control in their applications. This page highlights the control's flexibility in different scenarios and showcases potential real-world applications.
 <!-- 
 Please use this to aid the reviewer, this could include stating which platform(s) have been tested
 -->

### Platforms tested ###
- [x] Android
- [x] Windows
- [ ] MacCatalyst
- [ ] iOS
- [ ] Tizen 

 NOTE: Due to the developer not having access to iOS, MacCatalyst or Tizen, these platforms have not been tested or validated as working or visually correct.

GeorgeLeithead commented:
NOTE: I have NOT completed work on the C# syntax samples page, the documentation OR all of the Unit Tests!

pictos commented on src/CommunityToolkit.Maui/Views/RatingView/RatingView.shared.cs:
```diff
@@ -0,0 +1,627 @@
+			}
+			else if (i == fullShapes && partialFill > 0)
+			{
+				ratingShape.Fill = new LinearGradientBrush(
+					[
+							new GradientStop(filledColor, 0),
+							new GradientStop(filledColor, (float)partialFill),
+							new GradientStop(emptyColor, (float)partialFill),
+					],
+					new Point(0, 0), new Point(1, 0)); // Partial fill
```
this `LinearGradientBrush` construct, is almost the same as the one in at line 470, can we refactor it be a method? That way will be one code to maintain.

GeorgeLeithead replied:
Refactored out to a method.

pictos commented on src/CommunityToolkit.Maui/Views/RatingView/RatingView.shared.cs:
```diff
@@ -0,0 +1,627 @@
+namespace CommunityToolkit.Maui.Views;
+
+/// <summary>Rating view control.</summary>
+public class RatingView : TemplatedView, IRatingView
```
will we support negative values for RatingView?

GeorgeLeithead replied:
This is an excellent point, and I never thought about it.  I'll put in a Property validator and also a Clamp to make sure.

pictos reviewed (commented):
Great work! I've a few suggestions around the current implementation.

GeorgeLeithead commented on src/CommunityToolkit.Maui.Core/Interfaces/RatingView/IRatingViewShape.shared.cs:
```diff
@@ -0,0 +1,123 @@
+
+	/// <summary>Retrieves the default item shape size value.</summary>
+	/// <returns>The item shape <see cref="double"/> size.</returns>
+	double ItemShapeSizeDefaultValueCreator();
+
+	/// <summary>Retrieves the default shape value.</summary>
+	/// <returns>The shape <see cref="RatingViewShape"/>.</returns>
+	RatingViewShape ShapeDefaultValueCreator();
+
+	/// <summary>Retrieves the default shape border color value.</summary>
+	/// <returns>The item shape border <see cref="Color"/>.</returns>
+	Color ItemShapeBorderColorDefaultValueCreator();
+
+	/// <summary>Retrieves the default shape border thickness value.</summary>
+	/// <returns>The item shape border <see cref="double"/>.</returns>
+	double ItemShapeBorderThicknessDefaultValueCreator();
+
+	/// <summary>Retrieves the default rating item padding value.</summary>
+	/// <returns>The rating item padding <see cref="Thickness"/>.</returns>
+	Thickness ItemPaddingDefaultValueCreator();
+
+	/// <summary>Retrieves the default background color for an empty rating item.</summary>
+	/// <returns>The empty background color.</returns>
+	Color EmptyColorDefaultValueCreator();
+
+	/// <summary>Retrieves the default background color for an filled rating item.</summary>
+	/// <returns>The filled background color.</returns>
+	Color FilledColorDefaultValueCreator();
```
I was using the core .NET MAUI as a guide on how to implement, after all, if it's good enough for them. I have zero problem moving them into the main code base, if you think it's the right thing to do?

GeorgeLeithead commented on src/CommunityToolkit.Maui.Core/Primitives/RatingViewShape.cs:
```diff
@@ -0,0 +1,35 @@
+
+	/// <summary>Star shape.</summary>
+	/// <remarks>Default shape.</remarks>
+	public static RatingViewShape Star { get; } = new(PathShapes.Star);
```
This one, was mainly just a near direct copy from the original PR (that got stale with the original implementor).  I'm very appreciative of any suggestions and more than happy to take on any and all positive suggestions you can give.

GeorgeLeithead replied:
I'll see what I can do to implement what you suggested.

GeorgeLeithead replied:
Changed RatingViewShape to a sealed class with private constructor.

GeorgeLeithead commented on src/CommunityToolkit.Maui/Views/RatingView/RatingView.shared.cs:
```diff
@@ -0,0 +1,627 @@
+		return shapeBorder;
+	}
+
+	static List<VisualElement> GetVisualTreeDescendantsWithBorderAndShape(VisualElement root, bool isShapeFill)
```
Yes, I'll change this to be a read-only collection.

brminnick commented on src/CommunityToolkit.Maui/Views/RatingView/RatingView.shared.cs:
```diff
@@ -0,0 +1,748 @@
+	/// <summary>Bindable property for <see cref="ItemPadding"/> bindable property.</summary>
+	public static readonly BindableProperty ItemPaddingProperty = RatingViewItemElement.ItemPaddingProperty;
+
+	// TODO: Add some kind of Roslyn Analyser to validate at design/build time that the MaximumRating is in bounds (1-25).
```
Why is our maximum bound set to 25? Is this a platform-specific problem? 

Or are we just picking an arbitrary number for a maximum? 

I'm totally fine with us setting a maximum value, but I'm just curious if it's a platform constraint or an artificial constraint set by us?

brminnick reviewed (commented):
Thanks @GeorgeLeithead! I LOVE that you included a C# Markup Example. I've re-factored it to follow best practices:
- Flattened layouts (eg avoid putting a HorizontalLayout inside of a Grid)
- Use extension methods when available (e.g. `.Text()`, `CenterVertical` etc)
- Updated the Bindings
  - Use TypedBindings: https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/markup/extensions/bindable-object-extensions#bind
  - Use `[NotifyPropertyChangedFor]` in ViewModel
- Remove Unneeded `EventToCommandBehavior`
  - The previous solution worked, but added layers of complexity
  - Obsoleted by `[NotifyPropertyChangedFor]`
- Remove `Auto` from `RowDefinitions` + `ColumnDefinition`
  - `Auto` causes massive performance impacts because MAUI must recalculate every row in the entire Grid when one row's size increases when its `RowDefinition` is `Auto`
  
I haven't done a deep dive into the `RatingView` source code yet, but I did make some tweaks while updating the C# Markup Sample page:
- Rename `Shape` -> `ItemShape`
  - This helps ensure users know that they're setting the shape of the items in the RatingView to ensure they don't assume that `Shape` means the shape of the `RatingView` itself
- `SizeProperty` -> `ItemShapeSizeProperty`
  - This helps ensure users know that they're setting the size of the items in the RatingView to ensure they don't assume that `Size` means the size of the `RatingView` itself
- Change `RatingView.Control` from `public` -> `internal`
  - I don't understand `TemplatedView` super well, but I don't think the user will ever need to (or should) manually change this value
- Use `int` for `MaximumRating` and `Rating`
  - This helps ensure that users using intelligence understand that we expect and integer number for these values
  - `byte` allowed them to pass in a multitude of unsupported values 
- Add bounds to `ShapeBorderThickness`
  - This ensures we don't get negative values for border thickness
  - I added Unit Tests to support this scenario
- Add bounds to `MaximumRating`
  - This ensures we don't get invalid values for MaximumRating
  - I added Unit Tests to support this scenario
- Add bounds to `Rating`
  - This ensures that `Rating` never exceeds `MaximumRating`
  - I added Unit Tests to support this scenario
- Remove `static` modifier from `RatingView.weakEventManager`
  - When it was `static`, it would cause `event MaximumRating` to fire on all `RatingView` instances when `MaximumRating` changed on one instance
- Removed Null Coalescing Operator (`!`)
  - We need to always check for `null`; we should never assume an instance is not `null`
  - The `!` is a bit of an anti-pattern and I don't recommend any C# developer use it
- Removed `Debug.Assert`
  - Instead of using `Debug.Assert` to ensure an instance is not `null`, we should write the logic ourselves 
  - `Debug.Assert` doesn't run in RELEASE configuration (only in DEBUG configuration)
- Add `[NotNull]` to `EmptyColor`, `FilledColor`, and `ShapeBorderColor`
  - This allows users to pass in `null` for the value of these properties; we then check for `null` and change it `Colors.Transparant`
  - This will improve the XAML user experience where `Color` can default to `null`

GeorgeLeithead commented:
> * Use `int` for `MaximumRating` and `Rating`
>   
>   * This helps ensure that users using intelligence understand that we expect ad integer number for these values
>   * `byte` allowed them to pass in unsupported values

Not sure I agree here.  The `System.Byte` integral type is the smallest POSITIVE **unsigned 8-bit integer** available.  Whereas the `System.Int` integral type supports NEGATIVE and is a signed 32-bit integer, therefore the possible values are huge.  

Yes, I know we have added in `ValidateValue` handlers, but as an end user who will just see the data type of `int` may just assume it is the full signed 32-bit integer and offer negative and positive values.  Whereas, if they see it is a `byte`, this already draws the developer in to know that the value is a positive and restricted.

GeorgeLeithead commented:
> * Remove `static` modifier from `RatingView.weakEventManager`
>   
>   * When it was `static`, it would cause `event MaximumRating` to fire on all `RatingView` instances when `MaximumRating` changed on one instance

Thanks for this.  I wasn't sure why I was experiencing this in some early 'samples'.  I think it was mostly negated in the samples, when I added `IsReadOnly`.

GeorgeLeithead commented:
> Add bounds to `MaximumRating`
> * This ensures we don't get invalid values for MaximumRating
> * I added Unit Tests to support this scenario

Not sure I agree with this, and I feel that my original implementation was better.  In that values outside the range (1 to 25) would be set to the **nearest valid value**, and not throw errors.  Which is the SAME as [Opacity](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.visualelement.opacity?view=net-maui-8.0#microsoft-maui-controls-visualelement-opacity) and other common base properties.

GeorgeLeithead commented:
> Removed Null Coalescing Operator (`!`)
> * We need to always check for `null`; we should never assume an instance is not `null`
> * The `!` is a bit of an anti-pattern and I don't recommend any C# developer use it

How do we unit test these, as the `Code Coverage` report has now dropped from 100% to 85.9%?  **All branches** should be covered IMO.

GeorgeLeithead commented:
> Remove `Auto` from `RowDefinitions` + `ColumnDefinition`
> * `Auto` causes massive performance impacts because MAUI must recalculate every row in the entire Grid when one row's size increases when its `RowDefinition` is `Auto`

How does this handle when the user has their device set to anything other than the default "Font Size" and "Display size"?  How does this work for our visual impaired users?  Just increasing the "Display Size" by 1 setting, and most of the controls become un-usable! 
![image](https://github.com/user-attachments/assets/2760a33a-8968-473d-9c56-4244bd348b30)

brminnick commented:
> How do we unit test these, as the `Code Coverage` report has now dropped from 100% to 85.9%? **All branches** should be covered IMO, and if not covered are they needed?

As a general rule, if the Unit Test isn't covering a branch you've either written the library incorrectly or you need to add more unit tests. In this specific instance, our Unit Tests are still valid as long as we initialize `RatingView.Control`.

In the [Code Coverage report for this PR](https://dev.azure.com/dotnet/CommunityToolkit/_build/results?buildId=110280&view=codecoverage-tab), I see that our Unit Tests cover 94.27% of the lines in `RatingView` and 
100% of the lines in `RatingViewElement` (screen shot below). 

Can you help me understand where you're seeing that we only cover 85.9% of RatingView? 

<img width="1389" alt="Screenshot 2024-10-02 at 10 41 25 AM" src="https://github.com/user-attachments/assets/92d86f0f-0f99-4908-b7ee-f96002eb3586">

bijington commented on src/CommunityToolkit.Maui/Views/RatingView/RatingView.shared.cs:
```diff
@@ -0,0 +1,748 @@
+	}
+
+	/// <summary>Command that is triggered when the value of <see cref="Rating"/> is changed.</summary>
+	public ICommand? RatingChangedCommand
```
What's the thinking behind having a command here? Can't the value just be data bound to handle it changing?

bijington commented:
> > Removed Null Coalescing Operator (`!`)
> 
> > * We need to always check for `null`; we should never assume an instance is not `null`
> 
> > * The `!` is a bit of an anti-pattern and I don't recommend any C# developer use it
> 
> > Removed `Debug.Assert`
> 
> > * Instead of using `Debug.Assert` to ensure an instance is not `null`, we should write our own logic that helps explain why we're checking for null and our expected results
> 
> > * `Debug.Assert` doesn't run in RELEASE configuration (only in DEBUG configuration)
> 
> 
> 
> How do we unit test these, as the `Code Coverage` report has now dropped from 100% to 85.9%?  **All branches** should be covered IMO, and if not covered are they needed?
> 
> 
> 
> Is it even POSSIBLE to get to code such as this, and if not, is it even needed?
> 
> ```csharp
> 
> 		var ratingView = (RatingView)bindable;
> 
> 
> 
> 		if (ratingView.Control is null)
> 
> 		{
> 
> 			return;
> 
> 		}
> 
> ```

Could you achieve the passing of values to inner controls through bindings rather than doing it this way? Then you don't need to worry about checking for null

GeorgeLeithead commented:
> As a general rule, if the Unit Test isn't covering a branch you've either written the library incorrectly or you need to add more unit tests. In this specific instance, our Unit Tests are still valid as long as we initialize `RatingView.Control`.
> 
> In the [Code Coverage report for this PR](https://dev.azure.com/dotnet/CommunityToolkit/_build/results?buildId=110280&view=codecoverage-tab), I see that our Unit Tests cover 94.27% of the lines in `RatingView` and 100% of the lines in `RatingViewElement` (screen shot below).
> 
> Can you help me understand where you're seeing that we only cover 85.9% of RatingView?
> 
> <img alt="Screenshot 2024-10-02 at 10 41 25 AM" width="1389" src="https://private-user-images.githubusercontent.com/13558917/372974087-92d86f0f-0f99-4908-b7ee-f96002eb3586.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc5NTM0NjQsIm5iZiI6MTcyNzk1MzE2NCwicGF0aCI6Ii8xMzU1ODkxNy8zNzI5NzQwODctOTJkODZmMGYtMGY5OS00OTA4LWI3ZWUtZjk2MDAyZWIzNTg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAzVDEwNTkyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5NDUyMjJjMzAzNWIxMjViZTRkNTAzYTkwZGMwYTc4ZDVjYmFmYTJhN2UyYzQyNDE1NWNhZjYwODgwNjcwMWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BYt13ZoUEsgurj2QOOh6W4LRHM9K-ZZMlXWRVVsPKWM">

I was looking at the Branch Coverage.
![image](https://github.com/user-attachments/assets/cc9970b1-32cd-43c5-8f4e-6a7660f7ca63)
for example:
![image](https://github.com/user-attachments/assets/23062927-3584-4970-864d-e18adf83e80f)

I'll 'try' and see if I can get these covered.

GeorgeLeithead commented:
> > Remove `Auto` from `RowDefinitions` + `ColumnDefinition`
> > 
> > * `Auto` causes massive performance impacts because MAUI must recalculate every row in the entire Grid when one row's size increases when its `RowDefinition` is `Auto`
> 
> How does this handle when the user has their device set to anything other than the default "Font Size" and "Display size"? How does this work for our visual impaired users? Just increasing the "Display Size" by 1 setting, and most of the controls become un-usable! 

I have adjusted the row heights so that they display correctly with the device default "Font Size" and "Display Size" settings.

GeorgeLeithead commented:
@pictos I have added in a Rosalyn Analyzer to validate that the MaxaimumRating property is within the correct range.  Never written an analyzer before, so I hope it's good.

brminnick commented:
Hey George! I verified the Analyzer, but I'm unable to trigger the `MaximumRatingCodeFixProvider` (screen capture below). 

Are you able to confirm that the Code Fix for MaximumRating appears + works for you in Visual Studio? 
 
![ScreenFlow](https://github.com/user-attachments/assets/ff2447e7-b717-44e1-9812-ffb87f17d29b)

GeorgeLeithead commented:
> Hey George! I verified `MaximumRatingRangeAnalyzer` (and added another bounds-test and added Benchmarks for it), but I'm unable to trigger the `MaximumRatingCodeFixProvider` (screen capture below).
> 
> Are you able to confirm that the Code Fix for MaximumRating appears + works for you in Visual Studio? It should appear as the lightbulb offering to fix my code for me.
> 
> If you're unable to confirm `MaximumRatingCodeFixProvider` or unable to get it working, we can always just remove it because the error message in `MaximumRatingRangeAnalyzer` is very clear and is guaranteed to help the dev understand the constraints; the CodeFixProvider is a bit overkill, but certainly helpful if we get it working.
> 
> ![ScreenFlow](https://private-user-images.githubusercontent.com/13558917/399114994-ff2447e7-b717-44e1-9812-ffb87f17d29b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1Njc4NDAsIm5iZiI6MTczNTU2NzU0MCwicGF0aCI6Ii8xMzU1ODkxNy8zOTkxMTQ5OTQtZmYyNDQ3ZTctYjcxNy00NGUxLTk4MTItZmZiODdmMTdkMjliLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDE0MDU0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhNmZhZTdiYjk5NTk4NWE3NzliZWQ3MGI1NWNiODM2ZTFiMjYzYWUzNDJkMzlhOTIyZGQ0MzAzMzk3MmZhOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tgj2dsMwO8b49pyMbwfhhwcNKbkoJn_ZmujVSAQdI-w) [ ](https://private-user-images.githubusercontent.com/13558917/399114994-ff2447e7-b717-44e1-9812-ffb87f17d29b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1Njc4NDAsIm5iZiI6MTczNTU2NzU0MCwicGF0aCI6Ii8xMzU1ODkxNy8zOTkxMTQ5OTQtZmYyNDQ3ZTctYjcxNy00NGUxLTk4MTItZmZiODdmMTdkMjliLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDE0MDU0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhNmZhZTdiYjk5NTk4NWE3NzliZWQ3MGI1NWNiODM2ZTFiMjYzYWUzNDJkMzlhOTIyZGQ0MzAzMzk3MmZhOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tgj2dsMwO8b49pyMbwfhhwcNKbkoJn_ZmujVSAQdI-w) ![ScreenFlow](https://private-user-images.githubusercontent.com/13558917/399114994-ff2447e7-b717-44e1-9812-ffb87f17d29b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1Njc4NDAsIm5iZiI6MTczNTU2NzU0MCwicGF0aCI6Ii8xMzU1ODkxNy8zOTkxMTQ5OTQtZmYyNDQ3ZTctYjcxNy00NGUxLTk4MTItZmZiODdmMTdkMjliLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDE0MDU0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhNmZhZTdiYjk5NTk4NWE3NzliZWQ3MGI1NWNiODM2ZTFiMjYzYWUzNDJkMzlhOTIyZGQ0MzAzMzk3MmZhOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tgj2dsMwO8b49pyMbwfhhwcNKbkoJn_ZmujVSAQdI-w) [ ](https://private-user-images.githubusercontent.com/13558917/399114994-ff2447e7-b717-44e1-9812-ffb87f17d29b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1Njc4NDAsIm5iZiI6MTczNTU2NzU0MCwicGF0aCI6Ii8xMzU1ODkxNy8zOTkxMTQ5OTQtZmYyNDQ3ZTctYjcxNy00NGUxLTk4MTItZmZiODdmMTdkMjliLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDE0MDU0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhNmZhZTdiYjk5NTk4NWE3NzliZWQ3MGI1NWNiODM2ZTFiMjYzYWUzNDJkMzlhOTIyZGQ0MzAzMzk3MmZhOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tgj2dsMwO8b49pyMbwfhhwcNKbkoJn_ZmujVSAQdI-w)

I'm looking at it now.

GeorgeLeithead commented:
> I'm looking at it now.

@brminnick  got the Lightbulb showing correctly and offering to correct, but the code fix still needs some work.  Working on this...

GeorgeLeithead commented:
@brminnick The 'Lightbulb' shows correctly, but I just can't figure out why the ChangeDocument does not work!
![image](https://github.com/user-attachments/assets/691f3785-8e24-42e8-87c1-cc8e0615f45c)

GeorgeLeithead commented:
> @brminnick The 'Lightbulb' shows correctly, but I just can't figure out why the ChangeDocument does not work! ![image](https://private-user-images.githubusercontent.com/43666574/400834550-691f3785-8e24-42e8-87c1-cc8e0615f45c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYyNjg2MjIsIm5iZiI6MTczNjI2ODMyMiwicGF0aCI6Ii80MzY2NjU3NC80MDA4MzQ1NTAtNjkxZjM3ODUtOGUyNC00MmU4LTg3YzEtY2M4ZTA2MTVmNDVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA3VDE2NDUyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1Y2VlYjk5OWI4MGE4M2I1MGJmNDExZGQyY2VlZTZmM2ZlYzZhZWY2M2MyNDY3YTRmYzgyNWUwZDRmYTFiMjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.soYY2SiUPN97JgvO49KcHyL8iM0a0VIznEtKSTf457Q)

@pictos Would you have a suggestion on what I can do to make the `CodeFixProvider` work?

brminnick commented:
@GeorgeLeithead Let's just remove the CodeFix and keep the analyzer. 

The warning in the Analyzer is sufficient for developers to fix it themselves.

brminnick reviewed (changes requested):
Hey @GeorgeLeithead! I'm having trouble reconciling why we have both `public enum RatingViewShape` and `public sealed class RatingViewShape`. 

Do we need both? 
  - If yes, can we use different names for the types? 
  - If yes, can we move `public enum RatingViewShape` out of the file `/Interfaces/IRatingViewShape.shared.cs` and into its own file in the folder `/Primatives/[New File Name].shared.cs`

(Also adding this here: please remove `MaximumRatingAnalyzerCodeFixProvider` to avoid it blocking this PR since it's currently not working. You can totally add it back in a future PR, but I don't want this CodeFixProvider to be a blocker for `RatingView` since it's not critical for its release)

brminnick commented on src/CommunityToolkit.Maui.Core/Primitives/RatingViewShape.cs:
```diff
@@ -0,0 +1,41 @@
+﻿namespace CommunityToolkit.Maui.Core.Primitives;
```
Please fix the namespace here once we've decided how to reconcile `public enum RatingViewShape` + `public sealed class RatingViewShape` 

```suggestion
﻿namespace CommunityToolkit.Maui.Core;
```

brminnick commented:
Hey @GeorgeLeithead! Just following up - we just need to do two things to merge RatingView that I need your help with:

- Resolve the overlapping type names (`public enum RatingViewShape` vs `public sealed class RatingViewShape`)
- Update the Docs PR to reflect the naming changes

That's it! Then we can merge RatingView!!

GeorgeLeithead commented:
> Hey @GeorgeLeithead! Just following up - we just need to do two things to merge RatingView that I need your help with:
> 
> * Resolve the overlapping type names (`public enum RatingViewShape` vs `public sealed class RatingViewShape`)
> * Update the Docs PR to reflect the naming changes
> 
> That's it! Then we can merge RatingView!!

I'm looking at it now.

ne0rrmatrix commented on samples/CommunityToolkit.Maui.Sample/ViewModels/Views/RatingView/RatingViewViewModel.cs:
```diff
@@ -0,0 +1,58 @@
+
+namespace CommunityToolkit.Maui.Sample.ViewModels.Views;
+
+public partial class RatingViewXamlViewModel : BaseRatingViewViewModel;
```
It looks like you duplicated the line here?

TheCodeTraveler reviewed (dismissed):
Thanks @GeorgeLeithead!! RatingView is ready to go. Excellent work!! 

I've updated the namings a bit, and I'll submit a review now on the Docs PR that aligns with the updates 👍

FYI - the CI pipeline is currently blocked by a a compiler error in the Tizen build. We'll merge this PR as soon as @JoonghyunCho is able to help us fix the Tizen issue!
